Add GoogLeNet (Inception v1)

TheCodez opened the pull request:
This adds the GoogLeNet (Inception v1) model, including the auxiliary classifiers found in the paper. Related to Issue #537


I have updated the example project on my [branch](https://github.com/TheCodez/examples/tree/inception) to add support for training the _GoogLeNet_. Sadly I don't have the computing resources available to train the model on ImageNet.

TheCodez commented:
@fmassa what do you think?

TheCodez commented:
Those are the parameters used by the Berkely Vision GoogLeNet.
My [branch](https://github.com/TheCodez/examples/tree/inception) also includes the learning rate decay described in the paper.
https://github.com/TheCodez/examples/blob/inception/imagenet/README.md#googlenet

> python main.py -a googlenet --lr 0.01 --wd 0.0002 [imagenet-folder with train and val folders]

TheCodez commented:
@fmassa I just noticed, I missed a padding in `maxpool3` which resulted in the output size being `13x13x480` instead of the correct `14x14x480`.
Is there any difference in using `ceil_mode=True` vs. `padding=1` in the max pools?

TheCodez commented:
I have also added a batch normalized GoogLeNet version which could be trained using:

> python main.py -a googlenet_bn --lr 0.045 --wd 0.0002 [imagenet-folder with train and val folders]

I'm not entirely sure which learning rate decay strategy would be best suited, maybe the default one from the imagenet script.
Also for the original GoogLeNet, maybe the poly decay policy used by the Berkeley Vision GoogLeNet would be better suited than the one found in the paper.

fmassa commented:
Hi,

I just got back from holidays. The training that I launched didn't learn anything.

> Is there any difference in using ceil_mode=True vs. padding=1 in the max pools?

There is a difference between `ceil_mode=True` and `padding=1`, and they are complementary: `ceil_mode` corresponds to the rounding mode used (i.e., `ceil(input / stride)`, for the simplest example), while padding always add the specified values.

I can try running more training jobs, if you have the hyper-parameters to try.

TheCodez commented:
@fmassa the batch normalized version should be easier to train:

> python main.py -a googlenet_bn --lr 0.045 --wd 0.0002 [imagenet-folder with train and val folders]

For the regular GoogLeNet:

> python main.py -a googlenet --lr 0.01 --wd 0.0002 [imagenet-folder with train and val folders]

I have updated the training script adding some more data augmentation ColorJitter and LightingNoise, which are reduced in strength for the batch normalized version.
Also I'm now using the poly learning rate decay whereas the batch normalized version uses the default one. Script is here: https://github.com/TheCodez/examples/tree/inception/imagenet

Hopefully it succeds this time.

fmassa commented:
I launched another training with googlenet using your newer commits. Let's see what it gives

fmassa commented:
@TheCodez Doesn't look like it's learning anything on 8GPUs, with the defaults that you sent. It has run through 5 epochs already. I'll let it run overnight, but this seems to align with my expectations that reproducing their results might not be easy

TheCodez commented:
@fmassa 

The only obvious differences I notice to the Berkeley Vision version are:

- Using a padding of 1 in the max pool layers to achieve the right output sizes. Caffe uses `ceil` for rounding, so no padding is needed in that case. 
- Using the initialization scheme from tensorflow instead of "xavier".
- Using more advanced data augmentations, that are also used in the paper.

They needed 60 epochs (2,400,000 iterations) to achieve top-1 accuracy 68.7% (31.3% error) and a top-5 accuracy 88.9% (11.1% error) using the fast solver. My branch also includes the poly learning rate decay they use, maybe there is an error. I got the code for that from [here](https://github.com/meetshah1995/pytorch-semseg/commit/fb3136a9e065de7f7376b8287f49629faf3ad2e0#diff-e44f4a60e89e820dde9bb27afa634965R60). Im not really sure about [this line](https://github.com/TheCodez/examples/blob/inception/imagenet/main.py#L317) for calculating the current iteration: `iter = len(train_loader) * epoch + i`

Other than that my only idea is using a different learning rate to start with.

fmassa commented:
Hi @TheCodez ,

Here are my thoughts:
- we should try using `ceil=True` mode, instead of adding a padding
- we should use the exact same initialization from what they reported in the paper, this plays a big difference
- the data augmentations are not what is making the training not work at all

I believe the the fact is not training at all is probably a combination of weight initialization, data transformation (are their images in 0-1 or 0-255?). Can you double check that on top of existing implementations that are known to work?

TheCodez commented:
@fmassa I changed the weight initialization to `xavier` and added `ceil` mode. I used the initialization from the BVLC GoogLeNet as it is the only reprocucible version I found. I couldn't really tell which image ranges were used. Do you know what the default for Caffe is?

Also updated my training script to get rid off the line I wasn't sure about. https://github.com/TheCodez/examples/commit/dde173dca568a0b3b338ecc581cf3be040479c18

TheCodez commented:
@fmassa could you please test again, if you have time?

fmassa commented:
Hi @TheCodez ,

Sorry for the delay.

I didn't have the chance to try it again since last time. I think for googlenet, given that it seems to be hard to train from scratch, I think I'll end up providing only pre-trained versions (by reusing the original weights), and not enforce that they will have training code associated with it.

In this case, it would make sense to put it in torchhub, but I still need to figure out the details of how to do it.

TheCodez commented:
@fmassa my ideas would be:

- Train again with the updated code. According to https://github.com/soumith/imagenet-multiGPU.torch/issues/2#issuecomment-117268755 using the Xavier initialization, a torch version of the model was able to train correctly.

- Try the batch norm version, which should be a lot easier to  train. Maybe convert those weights for the non batch norm version or only provide the batch norm version.

- Convert weights from the tensorflow model, which slightly differs from the paper definiton. I think it also uses batch norm.

fmassa commented:
@TheCodez I'm currently leaning towards adapting the weights from TensorFlow, and providing it on the Hub.

TheCodez commented:
@fmassa I'll be trying to convert the weights from tensorflow.

TheCodez commented:
@fmassa I converted the weights. I haven't done extensive evaluation on the ImageNet validation set, instead I just tested some images and all of them were classified correctly.
I used a modified version of https://github.com/Cadene/tensorflow-model-zoo.torch to convert the weights, also had to do some modifications to the network itself to match the Tensorflow model.

The problem I'm currently having is that in Tensorflow the ImageNet dataset has _1001_ classes, which leads to me having to substract 1 from the prediction to get the correct classification.

@colesbury how did you solve that problem for the InceptionV3 model?

fmassa commented:
@TheCodez I think that we can remove the first row of the classifier weights, which should remove the first class and should be fine I think?

TheCodez commented:
@fmassa I've updated the code to match the structure required for the TensorFlow weights. Also added the input normalization used for the Inception v3 model.

Removing the first row did the trick. Weights are currently hosted here: https://github.com/TheCodez/vision/releases/tag/1.0

fmassa commented:
@TheCodez thanks for the update! I'm evaluating the pre-trained model on ImageNet to compare the results and I'll let you know

fmassa commented:
I got 
```
Acc@1 68.332 Acc@5 88.552
```
which seems about right, thanks! I'm uploading the pre-trained weights, and then I'll update the download path and the documentation, thanks!

TheCodez commented on torchvision/models/googlenet.py:
```diff
@@ -0,0 +1,183 @@
+        self.inception4c = Inception(512, 128, 128, 256, 24, 64, 64)
+        self.inception4d = Inception(512, 112, 144, 288, 32, 64, 64)
+        self.inception4e = Inception(528, 256, 160, 320, 32, 128, 128)
+        self.maxpool4 = nn.MaxPool2d(3, stride=2, ceil_mode=True)
```
@fmassa one thing to note here is that TensorFlow uses 2x2 pooling instead of 3x3. Don't know if that has a positive impact on the accuracy, but it would mean to further diverge from the paper.

fmassa replied:
What was the reason why you used 3x3 pooling, in order to make everything work out fine, given the differences between TF and PyTorch?

TheCodez replied:
This didn't cause problems during my conversion process so I probably just missed it. Should I change it? In that case it might be a good idea to add a note that the implementation differs from the paper.

fmassa replied:
let me try seeing if it makes a difference for the performance, and I'll let you know

fmassa replied:
Accuracy is 1 point better using the 2x2 pooling, with
```
Acc@1 69.778 Acc@5 89.530
```
so I'll be changing it. Thanks for the heads up

fmassa commented:
Thanks a lot @TheCodez !

TheCodez commented:
@fmassa Thank you for all the help and guidance 👍

fmassa commented:
Thanks for the PR! Keep the amazing work!
